frb_codegen/src/ir/file.rs

```diff
@@ -1,29 +1,123 @@
+use convert_case::{Case, Casing};
+
 use crate::target::Target;
-use crate::utils::misc::mod_from_rust_path;
+use crate::utils::misc::{is_multi_blocks_case, mod_from_rust_path, BlockIndex};
 use crate::{generator, ir::*, Opts};
 use std::collections::{HashMap, HashSet};
+use std::thread;
+
+use crate::utils::misc::ExtraTraitForVec;
 
 pub type IrStructPool = HashMap<String, IrStruct>;
 pub type IrEnumPool = HashMap<String, IrEnum>;
 
+use std::cell::RefCell;
+
+// this variable should be `none` for single block case, and not none for multi-blocks case.
+thread_local!(pub static SHARED_MODULE: RefCell<Option<String>> = RefCell::new(None));
+
+// shared types among all regular API blocks, in context of function paramters
+thread_local!(pub static SHARED_TYPES_INPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_INPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks in context of function outputs
+thread_local!(static SHARED_TYPES_OUTPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_OUTPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks
+thread_local!(static SHARED_TYPES_ALL: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_ALL: RefCell<bool> = RefCell::new(false));
+
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize))]
 pub struct IrFile {
-    pub funcs: Vec<IrFunc>,
+    /// for a regular API block: this field contains APIs defined in this block and
+    /// all methods of types(struct/enum) used ONLY in this block
+    /// ---NOTE: for shared types used in this block, their methods are also included here;
+    /// for a shared block: it ONLY contains methods from ALL shared types.
+    funcs: Vec<IrFunc>,
     pub struct_pool: IrStructPool,
     pub enum_pool: IrEnumPool,
     pub has_executor: bool,
+    pub block_index: BlockIndex,
+    pub shared: bool, // `true` for an auto-generated shared block, otherwise `false`
 }
 
 impl IrFile {
+    pub fn funcs(&self, exclude_shared_method: bool) -> Vec<IrFunc> {
+        if self.shared {
+            // assert each item of funcs are shared
+            for each in &self.funcs {
+                assert!(each.shared);
+            }
+            return self.funcs.clone();
+        }
+        match exclude_shared_method {
+            true => self
+                .funcs
+                .iter()
+                .cloned()
+                .filter(|each| !each.shared)
+                .collect::<Vec<_>>(),
+            false => self.funcs.clone(),
+        }
+    }
+
+    pub fn set_funcs(&mut self, dst_funcs: Vec<IrFunc>) {
+        self.funcs = dst_funcs;
+    }
+
+    pub fn append_funcs(&mut self, dst_funcs: Vec<IrFunc>) {
+        self.funcs.extend(dst_funcs);
+    }
```

Comment:
would be great to make IrFunc immutable

Reply:
I guess it is impossible to make the 2 funcs immutable.